src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -591,4 +591,86 @@ private function checkProductInStock(Product $product, AddProductToOrderCommand
+     * @param AddProductToOrderCommand $command
+     *
+     * @throws DuplicateProductInOrderException
+     */
+    private function assertProductDuplicate(Order $order, AddProductToOrderCommand $command): void
+    {
+        $invoicesContainingProduct = [];
+        foreach ($order->getOrderDetailList() as $orderDetail) {
+            if ($command->getProductId()->getValue() !== (int) $orderDetail['product_id']) {
+                continue;
+            }
+            if (!empty($command->getCombinationId()) && $command->getCombinationId() !== (int) $orderDetail['product_attribute_id']) {
+                continue;
+            }
+            $invoicesContainingProduct[] = (int) $orderDetail['id_order_invoice'];
+        }
+
+        if (empty($invoicesContainingProduct)) {
+            return;
+        }
+
+        // If it's a new invoice (or no invoice), the ID is null, so we check if the Order has invoice (in which case
+        // a new one is going to be created) If it doesn't have invoices we do'nt allow adding duplicate OrderDetail
+        if (empty($command->getOrderInvoiceId()) && !$order->hasInvoice()) {
+            throw new DuplicateProductInOrderException('You cannot add this product in the order has it is already present');
+        }
+
+        // If we are targeting a specific invoice check that the ID has not been found in the OrderDetail list
+        if (!empty($command->getOrderInvoiceId()) && in_array((int) $command->getOrderInvoiceId(), $invoicesContainingProduct)) {
+            throw new DuplicateProductInOrderException('You cannot add this product in the order has it is already present');
+        }
+    }
+
+    /**
+     * Update order details after a specific price has been created or updated
+     *
+     * @param Order $order
+     * @param Product $product
+     * @param int $productQuantity
+     * @param Number $priceTaxIncluded
+     * @param Number $priceTaxExcluded
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function updateOrderDetails(
+        Order $order,
+        Product $product,
+        int $combinationId,
+        int $productQuantity,
+        Number $priceTaxIncluded,
+        Number $priceTaxExcluded
+    ): void
+    {
+        foreach($order->getOrderDetailList() as $row) {
+            $orderDetail = new OrderDetail($row['id_order_detail']);
+            if ($orderDetail->product_id !== (int) $product->id) {
+                continue;
+            }
+            if (!empty($combinationId) && $combinationId !== (int) $orderDetail['product_attribute_id']) {
+                continue;
+            }
+            $orderDetail->unit_price_tax_excl = (float) (string) $priceTaxExcluded;
+            $orderDetail->unit_price_tax_incl = (float) (string) $priceTaxIncluded;
+            $orderDetail->total_price_tax_excl = Tools::ps_round((float) (string) $priceTaxExcluded * $productQuantity, $this->computingPrecision);
```

Comment:
You're totally right, but strangely I tested and the quantity remained correct even with this code 🤔 

This is really strange, I fixed it as you said anyway.

Reply:
About the condition, I can't use it because we're looping on orderDetails, `$row['id_order_detail']` and `$orderDetail->id` are always the same.

What I can do is pass the currently updated OrderDetail as a parameter to do this check 👍